### PR TITLE
Auth: add IsServiceAccount to IsRealUser

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**
This adds IsServiceAccount bool ontop of the SIgnedInUser struct to know whether the retrieved user is a service account or not.

**Why do we need this feature?**
We need to in order for us to differentiate between a retrieved user account and a service account account

**Which issue(s) does this PR fix?**:
https://github.com/grafana/grafana/issues/57532

**Special notes for your reviewer**:
Have checked with the original intent of IsRealUser and it is correct to assume that we want to exclude users whom are apikeys or service accounts.
IsRealUser is being used in these places https://github.com/grafana/grafana/search?q=IsRealUser

### CODE REVIEW

File: pkg/services/ngalert/schedule/schedule.go [hunk 1]
```diff
@@ -328,6 +328,7 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 		start := sch.clock.Now()
 
 		schedulerUser := &user.SignedInUser{
+			// FIXME: add is service account and refactor to a service account instead of a user
```
Comment (eleijonmarck) [inline pkg/services/ngalert/schedule/schedule.go:331]: @grafana/grafana-edge-squad wanted to add you into the loop to know how to account for grafana service accounts from the signedinuser and that we have added the `IsServiceAccount` attribute to the SignedInUser.

File: pkg/services/serviceaccounts/serviceaccounts.go [hunk 2]
```diff
@@ -7,13 +7,29 @@ import (
...
+Service accounts are used to authenticate API requests. They are not users and
+do not have a password.
+*/
 type Service interface {
 	CreateServiceAccount(ctx context.Context, orgID int64, saForm *CreateServiceAccountForm) (*ServiceAccountDTO, error)
 	DeleteServiceAccount(ctx context.Context, orgID, serviceAccountID int64) error
 	RetrieveServiceAccountIdByName(ctx context.Context, orgID int64, name string) (int64, error)
 }
 
+/*
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/serviceaccounts.go:22]: comments outside of the scope of this PR. this comment reflects on the migration of apikeys to service accounts

File: pkg/services/user/model.go [hunk 3]
```diff
@@ -203,6 +203,7 @@ type SignedInUser struct {
 	Name               string
 	Email              string
 	ApiKeyID           int64 `xorm:"api_key_id"`
+	IsServiceAccount   bool  `xorm:"is_service_account"`
```
Comment (eleijonmarck) [inline pkg/services/user/model.go:206]: bulk of the change is that we add this parameter

File: pkg/services/serviceaccounts/serviceaccounts.go [hunk 4]
```diff
@@ -7,13 +7,29 @@ import (
...
+
+migration from apikeys to service accounts:
+HideApiKeyTab is used to hide the api key tab in the UI.
+MigrateApiKeysToServiceAccounts migrates all API keys to service accounts.
+MigrateApiKey migrates a single API key to a service account.
+
+// only used for interal api calls
+RevertApiKey reverts a single service account to an API key.
+*/
 type Store interface {
```
Comment (kalleep) [inline pkg/services/serviceaccounts/serviceaccounts.go:33]: The store interface should not be exposed in the domain package. If we want to interact with service accounts the service should be injected. But I think that refactor should be done in another pr

File: pkg/services/user/model.go [hunk 5]
```diff
@@ -276,8 +277,13 @@ func (u *SignedInUser) HasRole(role roletype.RoleType) bool {
...
 }
 
+// IsRealUser returns true if the user is a real user and not a service account
 func (u *SignedInUser) IsRealUser() bool {
-	return u.UserID > 0
+	// backwards compatibility
+	// checking if userId the user is a real user
+	// previously we used to check if the UserId was 0 or -1
+	// and not a service account
+	return u.UserID > 0 && u.IsServiceAccount == false
```
Comment (kalleep) [inline pkg/services/user/model.go:286]: With this change we need to also update HasUniqueId to return true for service account, otherwise we will never cache permissions for a service account

**Review by kalleep — COMMENTED** [review]
Looks good, left a few comments we need to address

File: pkg/services/contexthandler/contexthandler.go [hunk 6]
```diff
@@ -312,7 +312,8 @@ func (h *ContextHandler) initContextWithAPIKey(reqContext *models.ReqContext) bo
 	}
 
 	if apikey.ServiceAccountId == nil || *apikey.ServiceAccountId < 1 { //There is no service account attached to the apikey
-		//Use the old APIkey method.  This provides backwards compatibility.
+		// Use the old APIkey method.  This provides backwards compatibility.
+		// will probably have to be supported for a long time.
```
Comment (toddtreece) [inline pkg/services/contexthandler/contexthandler.go:316]: > // will probably have to be supported for a long time.

i don't have a lot of context here, but are there downsides to making this migration automatic?
image

Comment (ryantxu) [reply pkg/services/contexthandler/contexthandler.go:316]: Or perhaps automatic if no keys exist

Comment (eleijonmarck) [reply pkg/services/contexthandler/contexthandler.go:316]: this is a good take. I'll add this to our backlog

Comment (eleijonmarck) [reply pkg/services/contexthandler/contexthandler.go:316]: @toddtreece alot of our customers rely on apikeys and we want to make sure that they have been aware of the change to service accounts.

Are you referring to a new instance in general, where they do not have any apikeys and we set service accounts tab only by default? 🤔

File: pkg/services/serviceaccounts/serviceaccounts.go (hunk 4 above)
Comment (eleijonmarck) [reply pkg/services/serviceaccounts/serviceaccounts.go:33]: il add a new issue for refactoring the store

File: pkg/services/contexthandler/contexthandler.go (hunk 6 above)
Comment (eleijonmarck) [reply pkg/services/contexthandler/contexthandler.go:316]: if it's the latter case, I totally agree that that does make sense ⚡

Comment (toddtreece) [reply pkg/services/contexthandler/contexthandler.go:316]: @eleijonmarck i guess i meant either way. i have no context about what actually happens behind the scenes when an API key is migrated using the `Migrate to service accounts now` button, i just see that a service account has been created and the API key still works as it did before migration:
image

and then the api keys have this notice:
image

Comment (eleijonmarck) [reply pkg/services/contexthandler/contexthandler.go:316]: added a new issue to the backlog - https://github.com/grafana/grafana/issues/58091

Comment (eleijonmarck) [reply pkg/services/contexthandler/contexthandler.go:316]: @toddtreece this is exactly the intended experience. We were the ones developing this workflow, so any feedback is welcome.
We could add some more detail about what is meant by migrating from apikey to service accounts.

Comment (toddtreece) [reply pkg/services/contexthandler/contexthandler.go:316]: nope, no issues with the workflow for migration, it works great! the `// will probably have to be supported for a long time` comment made me wonder if API keys could be automatically migrated at some point so that the logic for the old keys can be removed. a bit of a tangent since i was just curious about the plan, but that can be discussed outside of this PR :smile:

File: pkg/services/user/model.go [hunk 7]
```diff
@@ -276,16 +277,26 @@ func (u *SignedInUser) HasRole(role roletype.RoleType) bool {
...
 
+// IsServiceAccountUser returns true if the user is a service account
+func (u *SignedInUser) IsServiceAccountUser() bool {
+	return u.IsServiceAccount
+}
+
 func (u *SignedInUser) HasUniqueId() bool {
-	return u.IsRealUser() || u.IsApiKeyUser()
+	return u.IsRealUser() || u.IsApiKeyUser() || u.IsServiceAccountUser()
 }
```
Comment (eleijonmarck) [inline pkg/services/user/model.go:300]: @kalleep added the method IsServiceAccountUser() on top of the SignedUser struct to be able to use it the same way for apikeyUser and realUser.
This is to account for caching for the service accounts as well 👍

**Review by kalleep — APPROVED** [review]
